### PR TITLE
refactor(avoidance): generate shift line before the plan() function

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

- this PR contains small process flow update.
  - previously, shift lines are generated in plan().
  - in this PR, shift lines are generated in updateData() (before the plan() function)
- this PR also contains code cleaning.
  - define helper functions in header file.

## Test Perform

Evaluator Result

- [[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/8253c599-6048-519d-b643-1f571653b14f?project_id=prd_jt) This PR (817/825)
- [[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/981ab539-bd85-5a45-9fe2-781255a71554?project_id=prd_jt) Baseline (812/825)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -50,6 +50,22 @@ using tier4_autoware_utils::calcDistance2d;
 using tier4_autoware_utils::calcLateralDeviation;
 using tier4_planning_msgs::msg::AvoidanceDebugFactor;
 
+namespace
+{
+
+AvoidLine getNotStraightShiftLine(const AvoidLineArray & shift_lines)
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:56]: Non-straight sounds more natural
```suggestion
AvoidLine getNonStraightShiftLine(const AvoidLineArray & shift_lines)
```

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -2379,23 +2398,19 @@ void AvoidanceModule::addShiftLineIfApproved(const AvoidLineArray & shift_lines)
...
+    const auto sl_front = shift_lines.front();
+    const auto sl_back = shift_lines.back();
 
-    if (shift_lines.at(i).getRelativeLength() > 0.0) {
-      left_shift_array_.push_back({uuid_left_, shift_lines.front().start, shift_lines.back().end});
-    } else if (shift_lines.at(i).getRelativeLength() < 0.0) {
-      right_shift_array_.push_back(
-        {uuid_right_, shift_lines.front().start, shift_lines.back().end});
+    if (sl.getRelativeLength() > 0.0) {
+      left_shift_array_.push_back({uuid_left_, sl_front.start, sl_back.end});
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2411]: `emplace_back` is preferred here and next line. (This is not a part of PR, you don't need to change though)

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp [hunk 3]
```diff
@@ -200,6 +206,7 @@ class AvoidanceModule : public SceneModuleInterface
   void fillObjectMovingTime(ObjectData & object_data) const;
   void compensateDetectionLost(
     ObjectDataArray & target_objects, ObjectDataArray & other_objects) const;
+  void fillAvoidancePath(AvoidancePlanningData & data, DebugData & debug) const;
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp:209]: Is this function generating the avoidance path? I don't get a point of what "fill" means.

**Review by TakaHoribe — COMMENTED** [review]
@satoshi-ota Overall looks good to me. For my deep understanding, would you explain the background/reason behind this modification?

>this PR contains small process flow update.
previously, shift lines are generated in plan().
in this PR, shift lines are generated in updateData() (before the plan() function)

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp [hunk 4]
```diff
@@ -300,33 +306,91 @@ class AvoidanceModule : public SceneModuleInterface
   void updateAvoidanceDebugData(std::vector<AvoidanceDebugMsg> & avoidance_debug_msg_array) const;
   mutable std::vector<AvoidanceDebugMsg> debug_avoidance_initializer_for_shift_line_;
   mutable rclcpp::Time debug_avoidance_initializer_for_shift_line_time_;
-  // =====================================
+
   // ========= helper functions ==========
-  // =====================================
+
+  double getEgoSpeed() const
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp:312]: What is the reason for moving implementation to header?

Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp:312]: The number of lines in avoidance_module.cpp exceeds 3000. So I move simple helper fuction definition to header file in order to reduce the volume of cpp file.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: fix in https://github.com/autowarefoundation/autoware.universe/pull/2533/commits/980437c70ae3c43ed25684fbe868e019c5323d7c

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2411]: I'll fix in another PR. Thanks.

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp (hunk 4 above)
Comment (TakaHoribe) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp:312]: I got it. Or you can create another file avoidance_module_helper.cpp and put them there.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (TakaHoribe) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: Thanks!

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp (hunk 3 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp:209]: You're right :+1: 
I think `fillShiftLine` is better than `fillAvoidancePath`. What do you think ?

Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp]: fix in https://github.com/autowarefoundation/autoware.universe/pull/2533/commits/599a08cb891daa8283b045b68a47a9eb5d845618

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp (hunk 4 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp:312]: Nice idea :+1: I'll do that in refactoring ticket.

**Comment by satoshi-ota** [comment]
> @satoshi-ota Overall looks good to me. For my deep understanding, would you explain the background/reason behind this modification?
> 
> > this PR contains small process flow update.
> > previously, shift lines are generated in plan().
> > in this PR, shift lines are generated in updateData() (before the plan() function)

II'll add some processes about shift line generation in the other PRs, and I separate the process from `plan()` in order to avoid loss of readability.

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp (hunk 3 above)
Comment (TakaHoribe) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module.hpp]: Nice. Thank you

**Review by TakaHoribe — APPROVED** [review]

**Review by kminoda — APPROVED** [review]
LGTM

**Pull request merged**